File: 이진영/프로그래머스/오픈채팅방.js
function solution(record) {
    let uidMap = new Map();
    let recordArr = [];
    
    record.forEach(v => {
        const [type, uid, name] = v.split(" ");
        
        if (type === "Enter" || type === "Change") {
            uidMap.set(uid, name);
        }
        
        if (type === "Enter") {
            recordArr.push([1, uid]);
        } else if (type === "Leave") {
            recordArr.push([0, uid]);
        }
    })
    
    return recordArr.map(([type, uid]) => {
        if (type) {
            return `${uidMap.get(uid)}님이 들어왔습니다.`
        } else {
            return `${uidMap.get(uid)}님이 나갔습니다.`
        }
    })
}

File: 현세미/프로그래머스/오픈 채팅방.py
def solution(record):
    user_info = {}
    actions = []

    for log in record:
        words = log.split()
        action = words[0]
        user_id = words[1]

        if action == "Enter":
            nickname = words[2]
            user_info[user_id] = nickname
            actions.append((user_id, "님이 들어왔습니다."))
        elif action == "Leave":
            actions.append((user_id, "님이 나갔습니다."))
        elif action == "Change":
            nickname = words[2]
            user_info[user_id] = nickname

    result = []
    for action in actions:
        user_id, message = action
        nickname = user_info[user_id]
        result.append(nickname + message)

    return result
